src/libcore/num/dec2flt/mod.rs

```diff
@@ -0,0 +1,228 @@
+
+#![doc(hidden)]
+#![unstable(feature = "dec2flt",
+            reason = "internal routines only exposed for testing")]
+
+use prelude::*;
+use num::ParseFloatError as PFE;
+use num::FloatErrorKind;
+use self::parse::{parse_decimal, Decimal, Sign};
+use self::parse::ParseResult::{self, Valid, ShortcutToInf, ShortcutToZero};
+use self::num::digits_to_big;
+use self::rawfp::RawFloat;
+
+mod algorithm;
+mod table;
+mod num;
+// These two have their own tests.
+pub mod rawfp;
+pub mod parse;
+
+/// Entry point for decimal-to-f32 conversion.
+pub fn to_f32(s: &str) -> Result<f32, PFE> {
+    dec2flt(s)
+}
+
+/// Entry point for decimal-to-f64 conversion.
+pub fn to_f64(s: &str) -> Result<f64, PFE> {
+    dec2flt(s)
+}
+
+/// Split decimal string into sign and the rest, without inspecting or validating the rest.
+fn extract_sign(s: &str) -> (Sign, &str) {
+    match s.as_bytes()[0] {
+        b'+' => (Sign::Positive, &s[1..]),
+        b'-' => (Sign::Negative, &s[1..]),
+        // If the string is invalid, we never use the sign, so we don't need to validate here.
+        _ => (Sign::Positive, s),
+    }
+}
+
+/// Convert a decimal string into a floating point number.
+fn dec2flt<T: RawFloat>(s: &str) -> Result<T, PFE> {
+    if s.is_empty() {
+        return Err(PFE { __kind: FloatErrorKind::Empty });
+    }
+    let (sign, s) = extract_sign(s);
+    let flt = match parse_decimal(s) {
+        Valid(decimal) => try!(convert(decimal)),
+        ShortcutToInf => T::infinity(),
+        ShortcutToZero => T::zero(),
+        ParseResult::Invalid => match s {
+            "inf" => T::infinity(),
+            "NaN" => T::nan(),
+            _ => { return Err(PFE { __kind: FloatErrorKind::Invalid }); }
+        }
+    };
+
+    match sign {
+        Sign::Positive => Ok(flt),
+        Sign::Negative => Ok(-flt),
+    }
+}
+
+/// The main workhorse for the decimal-to-float conversion: Orchestrate all the preprocessing
+/// and figure out which algorithm should do the actual conversion.
+fn convert<T: RawFloat>(mut decimal: Decimal) -> Result<T, PFE> {
+    let mut e = decimal.exp;
+    simplify(&mut decimal, &mut e);
+    if let Some(x) = trivial_cases(&decimal, e) {
+        return Ok(x);
+    }
+    // AlgorithmM and AlgorithmR both compute approximately `f * 10^e`.
+    let max_digits = decimal.integral.len() + decimal.fractional.len() + e.abs() as usize;
+    // Remove/shift out the decimal point.
+    let e = e - decimal.fractional.len() as i64;
+    if let Some(x) = algorithm::fast_path(decimal.integral, decimal.fractional, e) {
+        return Ok(x);
+    }
+    // Big32x40 is limited to 1280 bits, which translates to about 385 decimal digits.
+    // If we exceed this, perhaps while calculating `f * 10^e` in Algorithm R or Algorithm M,
+    // we'll crash. So we error out before getting too close, with a generous safety margin.
+    if max_digits > 375 {
+        return Err(PFE { __kind: FloatErrorKind::Invalid });
```

Comment:
Thinking about it some more, I find it hard to believe that this is always strictly correct. Consider the exact decimal representation (this always exists and is terminating) of some float with an even mantissa. Add half an ULP. This still rounds to the same float as we started with because the mantissa was even. Now add 2^-10000 or some other sufficiently small power of two: Suddenly we need to round up, and it's not visible without inspecting thousands of additional bits!